File: backend/src/main/scala/domains/auth/objects/AuthPermissionFlags.scala
package domains.auth.objects

/** 账号权限标记，site manager 隐含题目和比赛管理权限。 */
final case class AuthPermissionFlags(
  siteManager: Boolean,
  problemManager: Boolean,
  contestManager: Boolean
)

/** 权限标记归一化工具，保证高权限包含低层管理能力。 */
object AuthPermissionFlags:
  /** 归一化权限布尔值；站点管理员会自动拥有 problem/contest manager。 */
  def normalize(siteManager: Boolean, problemManager: Boolean, contestManager: Boolean): AuthPermissionFlags =
    AuthPermissionFlags(
      siteManager = siteManager,
      problemManager = siteManager || problemManager,
      contestManager = siteManager || contestManager
    )

File: frontend/src/objects/auth/AuthPermissionFlags.ts
/** 账号站点权限位；siteManager 隐含题目和比赛管理权限。 */
export type AuthPermissionFlags = {
  siteManager: boolean
  problemManager: boolean
  contestManager: boolean
}

/** 归一化账号权限位；补齐 siteManager 对下级管理权限的隐含授权。 */
export function normalizeAuthPermissionFlags<T extends AuthPermissionFlags>(flags: T): T {
  return {
    ...flags,
    problemManager: flags.siteManager || flags.problemManager,
    contestManager: flags.siteManager || flags.contestManager,
  }
}
